### PR TITLE
Lesson on Scientific Debugging

### PR DESCRIPTION
**Pull request by droberts-ada**
## Description
This comes from the discussion Sockets had on Friday about debugging techniques, based on the section on evidence-based debugging at the end of the [Rails Debugging Activity Review](https://github.com/Ada-Developers-Academy/Teachers-Edition-Textbook/blob/master/00-programming-fundamentals/rails-debugging.md). Sockets really liked these ideas and requested that we introduce them earlier in the curriculum. I've heard that from previous cohorts as well, but now I've finally found the time to make it happen.

Makes the connection between debugging and the scientific method.

Targeted as a 1-hour lesson in week 5.

## Questions for the Author
- [ ] Have you updated the weekly learning goals in the pedagogy resource?
- [N/A] Have you updated the weekly assessment in Schoology?
- [x] Have you added links to the source (lucidchart, draw.io) for any diagram?
- [x] Have you updated the READMEs with links to new or moved resources?

## Todos
None

## Feedback Requested
- Is the example appropriate for where they'll be in the curriculum? My goal was to use something that would be fairly easy to understand, but I've underestimated the difficulty of code snippets in the past.
- There's a _lot_ here. I've tried to include plenty of exercises and questions, but I'm worried that it will feel like a giant wall of text, or that it will take way longer than an hour. Any thoughts on how we could break it up, reduce it, or otherwise improve the mechanics?

### CODE REVIEW

File: 00-programming-fundamentals/scientific-debugging.md [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+- Apply the scientific method to the debugging process
+- Define the term _repro_
+- Evaluate experiments in terms of _cost_ and _risk_
+- Use _controlled experiments_ to methodically investigate bugs
+
+## Introduction: How do you Debug?
+
+Many programmers, especially early in their careers, take an ad-hoc approach to debugging. This works fine for simple bugs, but is likely to fall apart as programs become more and more complex. The goal of this lesson is to define a process to make debugging efficient, repeatable and scalable.
+
+**Think/pair/share** (5/5/10 minutes): Spend some time reflecting on your debugging process. When you discover a bug in your code, how do you proceed? What steps do you follow to fix it?
```
Comment (tildeee) [inline 00-programming-fundamentals/scientific-debugging.md]: I'd like to give the students one or two concrete examples of bugs that they've likely debugged, just to get their minds working...

Consider:

**Think/pair/share** (5/5/10 minutes): Imagine a bug you see when something gets printed to the console twice when you expected it to be printed once. Spend some time reflecting on your debugging process. When you discover a bug in your code, how do you proceed? What steps do you follow to fix it?

Or consider adding a concrete example to the paragraph above?

**Comment by kaidamasaki** [comment]
To fix the wall of text issue I think this might want to be a few smaller lessons to help them digest this in chunks.

**Comment by kaidamasaki** [comment]
Potentially even having half the lesson, lunch, and then the second half of the lesson.

**Comment by droberts-ada** [comment]
@kaidamasaki I like that idea but I'm not sure I see a clean way to break it up. Do you have something in mind?

**Comment by droberts-ada** [comment]
Whoops, missed your second post. Yeah, that's not a bad idea. I do think that trying to do this whole lesson in an hour is a bit of a stretch.

**Comment by tildeee** [comment]
Oh! To respond to a different thing: I'd love this to be either Hotel week or earlier.

**Pull request merged**